typescript/sdk/src/ism/HyperlaneIsmFactory.ts

```diff
@@ -0,0 +1,250 @@
+          chain,
+          aggregationIsmFactory,
+          addresses,
+          config.threshold,
+        );
+        return StaticAggregationIsm__factory.connect(address, signer);
+      }
+      default: {
+        throw new Error('Unknown ModuleType');
+      }
+    }
+  }
+
+  private async deployMOfNFactory(
+    chain: ChainName,
+    factory: StaticMOfNAddressSetFactory,
+    values: types.Address[],
+    threshold: number,
+  ): Promise<types.Address> {
+    const address = await factory.getAddress(values.sort(), threshold);
+    const provider = this.multiProvider.getProvider(chain);
+    const code = await provider.getCode(address);
+    if (code === '0x') {
+      await factory.deploy(values.sort(), threshold);
+    }
+    return address;
+  }
+}
+
+export async function moduleMatches(
+  chain: ChainName,
+  moduleAddress: types.Address,
+  config: IsmConfig,
+  multiProvider: MultiProvider,
+  contracts: HyperlaneContracts<IsmFactoryFactories>,
+): Promise<boolean> {
+  const provider = multiProvider.getProvider(chain);
+  const module = IInterchainSecurityModule__factory.connect(
+    moduleAddress,
+    provider,
+  );
+  const actualType = await module.moduleType();
+  if (actualType !== config.type) return false;
+  let matches = true;
+  switch (config.type) {
+    case ModuleType.MULTISIG: {
+      // A MultisigIsm matches if validators and threshold match the config
+      const expectedAdddress = await contracts.multisigIsmFactory.getAddress(
+        config.validators.sort(),
+        config.threshold,
+      );
+      matches = eqAddress(expectedAdddress, module.address);
+      break;
+    }
+    case ModuleType.ROUTING: {
+      // A RoutingIsm matches if:
+      //   1. The set of domains in the config equals those on-chain
+      //   2. The modules for each domain match the config
+      // TODO: Check (1)
+      const routingIsm = DomainRoutingIsm__factory.connect(
+        moduleAddress,
+        provider,
+      );
+      const owner = await routingIsm.owner();
+      matches = matches && eqAddress(owner, config.owner);
+      for (const chain of Object.keys(config.domains)) {
+        const subModule = await routingIsm.modules(
+          multiProvider.getDomainId(chain),
+        );
+        const subModuleMatches = await moduleMatches(
+          chain,
+          subModule,
+          config.domains[chain],
+          multiProvider,
+          contracts,
+        );
+        matches = matches && subModuleMatches;
+      }
+      break;
+    }
+    case ModuleType.AGGREGATION: {
+      // An AggregationIsm matches if:
+      //   1. The threshold matches the config
+      //   2. There is a bijection between on and off-chain configured modules
+      const aggregationIsm = StaticAggregationIsm__factory.connect(
+        moduleAddress,
+        provider,
+      );
+      const [modules, threshold] = await aggregationIsm.modulesAndThreshold(
+        '0x',
+      );
```

Comment:
how do we validate a match if message does influence returned values?

Reply:
Great question, as far as I can tell we cannot. This function assumes specific implementations of these ISM interfaces. Not sure how useful that is.